syntax = "proto3";

import "types.proto";

package pairing_server;

/* XXX: Need to handle more general searches, like get all players in
 * tournament and such! */

message Request {
    enum Operation {
        PING = 0;
        TOURNAMENT_CREATE = 1;
        TOURNAMENT_GET = 2;
        TOURNAMENT_EDIT = 3;
        PLAYER_SIGNUP = 4;
        PLAYER_GET = 5;
        PLAYER_EDIT = 6;
        GAME_GET = 7;
        GAME_REGISTER_RESULT = 8;

        TOURNAMENT_GET_ROUND = 9;
        TOURNAMENT_GET_PLAYERS = 10;
    }
    Tournament tournament = 1;
    TournamentPlayer player = 2;
    TournamentGame game = 3;
    Operation operation = 4;
}

message Response {
    enum Status {
        OK = 0;
        ERROR = 1;
    }
    enum Reason {
        UNSPECIFIED = 0;
        INTERNAL_ERROR = 1;
        MISSING_HMAC = 2;
        WRONG_HMAC = 3;
        MISSING_UUID = 4;
        NOT_FOUND = 5;
        INCOMPLETE = 6;
        /* TODO: All the other reasons things can go wrong. */
    }
    Status status = 1;
    repeated Tournament tournaments = 2;
    repeated TournamentPlayer players = 3;
    repeated TournamentGame games = 4;
    repeated Reason reasons = 5;
}
